okhttp/src/main/java/okhttp3/internal/ws/RealWebSocket.java

```diff
@@ -52,38 +55,37 @@
   /** True after calling {@link #close()} to free connection resources. */
   private final AtomicBoolean connectionClosed = new AtomicBoolean();
 
-  public RealWebSocket(boolean isClient, BufferedSource source, BufferedSink sink, Random random,
-      final Executor replyExecutor, final WebSocketListener listener, final String url) {
+  protected RealWebSocket(boolean isClient, BufferedSource source, BufferedSink sink, Random random,
+      Executor replyExecutor, WebSocketListener listener, String url) {
+    this.replyExecutor = replyExecutor;
     this.listener = listener;
+    this.url = url;
 
     writer = new WebSocketWriter(isClient, sink, random);
-    reader = new WebSocketReader(isClient, source, new FrameCallback() {
-      @Override public void onMessage(ResponseBody message) throws IOException {
-        listener.onMessage(message);
-      }
+    reader = new WebSocketReader(isClient, source, this);
```

Comment:
Here's the allocation elimination. Moved the listener to just be on the enclosing type.